File: test/watershed/tree/client_io.mjs
import { readFileSync } from "node:fs";
import { createInterface } from "node:readline";

function startup(code, operation, message) {
  process.stderr.write(`${JSON.stringify({
    kind: "startup-error",
    code,
    operation,
    message,
  })}\n`);
  process.exit(1);
}

export function descriptor() {
  if (!process.env.WATERSHED_DESCRIPTOR) {
    startup("descriptor-decode-failed", "descriptor", "Missing WATERSHED_DESCRIPTOR");
  }
  try {
    return readFileSync(process.env.WATERSHED_DESCRIPTOR, "utf8");
  } catch (error) {
    startup("descriptor-decode-failed", "descriptor", error.message);
  }
}

export function token() {
  if (!process.env.WATERSHED_TOKEN) {
    startup("bootstrap-failed", "connect", "Missing WATERSHED_TOKEN");
  }
  return process.env.WATERSHED_TOKEN;
}

export function wait(milliseconds) {
  return new Promise((resolve) => setTimeout(resolve, milliseconds));
}

export async function lines(handle, finish) {
  try {
    for await (const line of createInterface({ input: process.stdin, crlfDelay: Infinity })) {
      const reply = await handle(line);
      process.stdout.write(`${reply}\n`);
    }
  } finally {
    finish();
  }
  await new Promise((resolve) => process.stdout.write("", resolve));
  process.exit(0);
}

export function fail(message) {
  process.stderr.write(`${message}\n`);
  process.exitCode = 1;
}

File: test/watershed/tree/client_io.erl
-module(client_io).
-export([descriptor/0, token/0, read_line/0, write_line/1, fail/1]).

descriptor() ->
    case os:getenv("WATERSHED_DESCRIPTOR") of
        false ->
            startup("descriptor-decode-failed", "descriptor",
                "Missing WATERSHED_DESCRIPTOR");
        Path ->
            case file:read_file(Path) of
                {ok, Bytes} -> Bytes;
                {error, Reason} ->
                    startup("descriptor-decode-failed", "descriptor",
                        ["Could not read descriptor: ", atom_to_list(Reason)])
            end
    end.

token() ->
    case os:getenv("WATERSHED_TOKEN") of
        false -> startup("bootstrap-failed", "connect", "Missing WATERSHED_TOKEN");
        Value -> unicode:characters_to_binary(Value)
    end.

read_line() ->
    case io:get_line("") of
        eof -> {error, nil};
        {error, Reason} -> erlang:error({stdin_error, Reason});
        Line -> {ok, unicode:characters_to_binary(Line)}
    end.

write_line(Value) ->
    ok = io:put_chars([Value, "\n"]),
    nil.

fail(Value) ->
    io:format(standard_error, "~ts~n", [Value]),
    erlang:halt(1).

startup(Code, Operation, Message) ->
    Encoded = io_lib:format(
        "{\"kind\":\"startup-error\",\"code\":\"~ts\",\"operation\":\"~ts\",\"message\":\"~ts\"}",
        [escape(Code), escape(Operation), escape(Message)]
    ),
    fail(Encoded).

escape(Value) ->
    Bytes = unicode:characters_to_binary(Value),
    EscapedSlash = binary:replace(Bytes, <<"\\">>, <<"\\\\">>, [global]),
    EscapedQuote = binary:replace(EscapedSlash, <<"\"">>, <<"\\\"">>, [global]),
    EscapedReturn = binary:replace(EscapedQuote, <<"\r">>, <<"\\r">>, [global]),
    binary:replace(EscapedReturn, <<"\n">>, <<"\\n">>, [global]).
